[quant][graph] Add a new observer type for dynamic quantization

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35366 [quant][graph] Add pass to insert quant dequant for dynamic quantization
* #35325 [quant][graph] Update dynamic quant tests to use new qconfig
* **#35265 [quant][graph] Add a new observer type for dynamic quantization**
* #35235 [quant][graph] Add _choose_qparams function for graph mode

Summary:
In graph mode we need to observer the activation tensor for dynamic quantization. This observer should behave the same way as the quantization functions called in the dynamic operator.
Currently for qlinear_dynamic we call quant_utils::ChooseQuantizationParams which has its own logic for calculating scale and zero_point.
We mimic those calculations in the new observer.

Test Plan:
python test/test_quantization.py ObserverTest

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20630988](https://our.internmc.facebook.com/intern/diff/D20630988)

supriyar committed:
[quant][graph] Add a new observer type for dynamic quantization
Summary:
In graph mode we need to observer the activation tensor for dynamic quantization. This observer should behave the same way as the quantization functions called in the dynamic operator.
Currently for qlinear_dynamic we call quant_utils::ChooseQuantizationParams which has its own logic for calculating scale and zero_point.
We mimic those calculations in the new observer.

Test Plan:
python test/test_quantization.py ObserverTest

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented on torch/quantization/observer.py:
```diff
@@ -439,6 +439,78 @@ def forward(self, x_orig):
+        )
+
+        if self.dtype == torch.qint8:
+            if self.reduce_range:
```
@raghuramank100 I didn't change `reduce_range` in this PR as it also part of MinMaxObserver and `_choose_qparams_per_tensor`. I thought we could have a separate change where we update reduce_range everywhere to `qmin` and `qmax`

raghuramank100 replied:
I have a PR for that that I need to update at: https://github.com/pytorch/pytorch/pull/33743

supriyar committed:
Update on "[quant][graph] Add a new observer type for dynamic quantization"


Summary:
In graph mode we need to observer the activation tensor for dynamic quantization. This observer should behave the same way as the quantization functions called in the dynamic operator.
Currently for qlinear_dynamic we call quant_utils::ChooseQuantizationParams which has its own logic for calculating scale and zero_point.
We mimic those calculations in the new observer.

Test Plan:
python test/test_quantization.py ObserverTest

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add a new observer type for dynamic quantization"


Summary:
In graph mode we need to observer the activation tensor for dynamic quantization. This observer should behave the same way as the quantization functions called in the dynamic operator.
Currently for qlinear_dynamic we call quant_utils::ChooseQuantizationParams which has its own logic for calculating scale and zero_point.
We mimic those calculations in the new observer.

Test Plan:
python test/test_quantization.py ObserverTest

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add a new observer type for dynamic quantization"


Summary:
In graph mode we need to observer the activation tensor for dynamic quantization. This observer should behave the same way as the quantization functions called in the dynamic operator.
Currently for qlinear_dynamic we call quant_utils::ChooseQuantizationParams which has its own logic for calculating scale and zero_point.
We mimic those calculations in the new observer.

Test Plan:
python test/test_quantization.py ObserverTest

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add a new observer type for dynamic quantization"


Summary:
In graph mode we need to observer the activation tensor for dynamic quantization. This observer should behave the same way as the quantization functions called in the dynamic operator.
Currently for qlinear_dynamic we call quant_utils::ChooseQuantizationParams which has its own logic for calculating scale and zero_point.
We mimic those calculations in the new observer.

Test Plan:
python test/test_quantization.py ObserverTest

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on test/test_quantization.py:
```diff
@@ -1336,6 +1337,25 @@ def test_per_tensor_observers(self, qdtype, qscheme, reduce_range):
+        qparams = obs.calculate_qparams()
+        ref = torch._choose_qparams_per_tensor(x, reduce_range)
+
+        np.testing.assert_array_almost_equal(ref[0], qparams[0], decimal=4)
```
Can we make this match exact?

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -439,6 +439,78 @@ def forward(self, x_orig):
+        min_val = min(0.0, min_val)
+        max_val = max(0.0, max_val)
+        scale = (max_val - min_val) / float(qmax - qmin)
+        if scale == 0.0:
```
Do you also need to check for large values of scale to match with the c++ implementation?

supriyar committed:
Update on "[quant][graph] Add a new observer type for dynamic quantization"


Summary:
In graph mode we need to observer the activation tensor for dynamic quantization. This observer should behave the same way as the quantization functions called in the dynamic operator.
Currently for qlinear_dynamic we call quant_utils::ChooseQuantizationParams which has its own logic for calculating scale and zero_point.
We mimic those calculations in the new observer.

Test Plan:
python test/test_quantization.py ObserverTest

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20630988](https://our.internmc.facebook.com/intern/diff/D20630988)

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add a new observer type for dynamic quantization"


Summary:
In graph mode we need to observer the activation tensor for dynamic quantization. This observer should behave the same way as the quantization functions called in the dynamic operator.
Currently for qlinear_dynamic we call quant_utils::ChooseQuantizationParams which has its own logic for calculating scale and zero_point.
We mimic those calculations in the new observer.

Test Plan:
python test/test_quantization.py ObserverTest

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20630988](https://our.internmc.facebook.com/intern/diff/D20630988)

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add a new observer type for dynamic quantization"


Summary:
In graph mode we need to observer the activation tensor for dynamic quantization. This observer should behave the same way as the quantization functions called in the dynamic operator.
Currently for qlinear_dynamic we call quant_utils::ChooseQuantizationParams which has its own logic for calculating scale and zero_point.
We mimic those calculations in the new observer.

Test Plan:
python test/test_quantization.py ObserverTest

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20630988](https://our.internmc.facebook.com/intern/diff/D20630988)

[ghstack-poisoned]

supriyar committed:
Update on "[quant][graph] Add a new observer type for dynamic quantization"


Summary:
In graph mode we need to observer the activation tensor for dynamic quantization. This observer should behave the same way as the quantization functions called in the dynamic operator.
Currently for qlinear_dynamic we call quant_utils::ChooseQuantizationParams which has its own logic for calculating scale and zero_point.
We mimic those calculations in the new observer.

Test Plan:
python test/test_quantization.py ObserverTest

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D20630988](https://our.internmc.facebook.com/intern/diff/D20630988)

[ghstack-poisoned]
